chore: add docs on rewrites and redirects (#767)
netlify/netlify-plugin-nextjs: docs/redirects-rewrites.md

# Redirects and rewrites
Version 4 of the Essential Next.js build plugin adds support for native Next.js [rewrites](https://nextjs.org/docs/api-reference/next.config.js/rewrites) and [redirects](https://nextjs.org/docs/api-reference/next.config.js/redirects). These are defined in your `next.config.js` file and include support for some features that are not included in Netlify redirects and rewrites.

## Using Netlify redirects and rewrites on a Next.js site

> Every site on Netlify supports [redirects and rewrites](https://docs.netlify.com/routing/redirects/), which are defined in a `_redirects` file or `netlify.toml`, and sites that use this plugin are no exceptions. However there are some caveats to bear in mind when using them. The plugin generates several rewrites of its own, which are used to map paths from the site to different Netlify functions which handle SSR, preview mode and images, as well as assets in `/_next/static`. Any Netlify redirects or rewrites that you create take precedence over these rewrites, so you should avoid adding a root level rewrite, because that would override the rewrites required by the plugin.

Reviewer: [dust] Having an example here might be helpful, but super clear writing 👌

> Every site on Netlify supports [redirects and rewrites](https://docs.netlify.com/routing/redirects/), which are defined in a `_redirects` file or `netlify.toml`, and sites that use this plugin are no exceptions. However there are some caveats to bear in mind when using them. The plugin generates several rewrites of its own, which are used to map paths from the site to different Netlify functions which handle SSR, preview mode and images, as well as assets in `/_next/static`. Any Netlify redirects or rewrites that you create take precedence over these rewrites, so you should avoid adding a root level rewrite, because that would override the rewrites required by the plugin.

Reviewer: ```suggestion
Every site on Netlify supports [redirects and rewrites](https://docs.netlify.com/routing/redirects/), which are defined in a `_redirects` file or `netlify.toml`, and sites that use this plugin are no exceptions. However there are some caveats to bear in mind when using them. The plugin generates several rewrites of its own, which are used to map paths from the site to different Netlify functions which handle SSR, preview mode and images, as well as assets in `/_next/static`. Any Netlify redirects or rewrites that you create [take precedence over these rewrites](#Redirect-and-rewrite-precedence), so you should avoid adding a root level rewrite, because that would override the rewrites required by the plugin.
```

> Every site on Netlify supports [redirects and rewrites](https://docs.netlify.com/routing/redirects/), which are defined in a `_redirects` file or `netlify.toml`, and sites that use this plugin are no exceptions. However there are some caveats to bear in mind when using them. The plugin generates several rewrites of its own, which are used to map paths from the site to different Netlify functions which handle SSR, preview mode and images, as well as assets in `/_next/static`. Any Netlify redirects or rewrites that you create [take precedence over these rewrites](#Redirect-and-rewrite-precedence), so you should avoid adding a root level rewrite, because that would override the rewrites required by the plugin.

Reviewer 2: I'm late to the party on this one after being out for a few days, but an example of what you mean by a root level rewrite could be useful. I may be off base, but something like this? 

```
/*   /index.html   200
```

If you make this addition, this sentence could benefit from one more `Netlify` mention to be crystal clear:

> Any Netlify redirects or rewrites that you create [take precedence over these rewrites](#Redirect-and-rewrite-precedence), so you should avoid adding a root level Netlify rewrite, because that would override the rewrites required by the plugin.

## Redirect and rewrite precedence
Rewrites and redirects are applied in the following order:

1. Redirects and rewrites in the `_redirects` file. These are read in order until a match is found, and then processing stops.
2. Redirects and rewrites in the `netlify.toml` file. None of these are read if one previous rules has already matched.
3. At this point, if the request targets a static file then it will be returned, without the Next.js redirects or rewrites being evaluated.

> 4. Any request that does not target a static file will then be passed to Next.js, which will then evaluate redirects and rewrites. These are defined in the `next.config.js` file.

Reviewer: Love this section!

The part with **These are defined in the `next.config.js` file.**, it's not super clear what `These` refers to. I'm guessing you're referring to the redirects and rewrites?

```suggestion
4. Any request that does not target a static file will then be passed to Next.js, and then evaluate redirects and rewrites which are defined in the `next.config.js` file.
```

## General principles

Netlify and Next.js redirects support different features and are evaluated at different points in the request lifecycle, so there are some considerations in deciding which one to use with your site.

### When to use Netlify redirects or rewrites:

> - Generally if your redirect can be handled with Netlify redirects, these are faster to evaluate and should be preferred.

Reviewer: [sand] I'm guessing we should have some metrics/blog posts showing that Netlify is faster at some point.

> - Features such as identity, proxying and country-based redirects are Netlify-specific and must use Netlify redirects.

Reviewer: Are there more features? If no then to lessen any possible ambiguity, something like:

```suggestion
- [Identity](https://docs.netlify.com/visitor-access/identity/), [proxying](https://docs.netlify.com/routing/redirects/rewrites-proxies/) and [country-based redirects](https://docs.netlify.com/routing/redirects/) are Netlify-specific features and must use Netlify redirects.
```
